### ISSUE TITLE
mont_mult_* calling square_w_32 violates type-based alias rules

### ISSUE THREAD

**Issue by rguenth** [Legrandin/pycryptodome#291]
On 32bit platforms at least it seems that we have

size_t inline square_w(uint64_t *t, const uint64_t *a, size_t words)
{   
    return square_w_32((uint32_t*)t, (const uint32_t*)a, words*2)/2;
}

and square_w_32 reading/writing to t via uint32_t.  That violates type-based alias rules
for example at the caller mont_mult_p256 which does

    if (a == b) {
        square_w(t, a, 4U);
    } else {
        product(t, a, b, 4U);
    }   
        
    t[(2*4U +1)-1] = 0; 

    for (i=0; i<(2*4U +1); i++) {
        t32[2*i] = (uint32_t)t[i];
        t32[2*i+1] = (uint32_t)(t[i] >> 32);
    }

and in this loop reads back values via uint64_t types (t is uint64_t[]).  This causes
GCC to happily optimize the first read (from t[0]) to zero because it can see
the memset to zero in square_w bypassing the stores via uint32_t.

Using the uint64_t scratchpad for uint32_t values is OK but store and load types
have to match.

A way out is to use -fno-strict-aliasing or declaring aliasing types via
typedef uint32_t aliasing_uint32_t __attribute__((may_alias));
and using aliasing_uint32_t * for the types in the affected functions.

There may be more issues in the code base like addmul32 and friends.

**Comment by marxin** [comment]
Note that the issue was seen after update to `GCC 9.1.0` and it fails on `i586` in tests like this:

```
Traceback (most recent call last):
  File "setup.py", line 493, in <module>
    ext_modules=ext_modules,
  File "/usr/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 186, in run
    SelfTest.run(module=moduleObj, verbosity=self.verbose, stream=sys.stdout, config=self.config)
  File "build/lib.linux-i686-3.7/Crypto/SelfTest/__init__.py", line 61, in run
    tests = get_tests(config=config)
  File "build/lib.linux-i686-3.7/Crypto/SelfTest/__init__.py", line 85, in get_tests
    from Crypto.SelfTest import PublicKey; tests += PublicKey.get_tests(config=config)
  File "build/lib.linux-i686-3.7/Crypto/SelfTest/PublicKey/__init__.py", line 35, in get_tests
    from Crypto.SelfTest.PublicKey import test_ECC;       tests += test_ECC.get_tests(config=config)
  File "build/lib.linux-i686-3.7/Crypto/SelfTest/PublicKey/test_ECC.py", line 36, in <module>
    from Crypto.PublicKey import ECC
  File "build/lib.linux-i686-3.7/Crypto/PublicKey/ECC.py", line 380, in <module>
    p256_G = EccPoint(_curves['p256'].Gx, _curves['p256'].Gy, "p256")
  File "build/lib.linux-i686-3.7/Crypto/PublicKey/ECC.py", line 258, in __init__
    raise ValueError("The EC point does not belong to the curve")
```

**Comment by marxin** [comment]
And the issue started with 0b92946ee4a53875033798e6b590e1b27e30c366.

**Comment by Legrandin** [comment]
Thanks for reporting. This is fixed in v3.8.2.

**Comment by marxin** [comment]
Thank you for the fix. I can confirm it works with `GCC 9.1.0.`
